volume/drivers/nfs/nfs.go

```diff
@@ -285,31 +294,131 @@ func (d *driver) Create(
+	if spec.GetShared() {
+		// File based
+		v = common.NewVolume(
+			volumeID,
+			api.FSType_FS_TYPE_NFS,
+			locator,
+			source,
+			spec,
+		)
+
+		// Create a directory on the NFS server with this UUID.
+		err := os.MkdirAll(volPath, 0744)
+		if err != nil {
+			logrus.Println(err)
+			return "", err
+		}
 
-	if err := f.Truncate(int64(spec.Size)); err != nil {
-		logrus.Println(err)
-		return "", err
-	}
+		// Setup volume
+		v.State = api.VolumeState_VOLUME_STATE_PENDING
+		if err := d.CreateVol(v); err != nil {
+			return "", err
+		}
 
-	v := common.NewVolume(
-		volumeID,
-		api.FSType_FS_TYPE_NFS,
-		locator,
-		source,
-		spec,
-	)
-	v.DevicePath = path.Join(volPathParent, volumeID+nfsBlockFile)
+		// Check for cloning
+		if source != nil && len(source.GetParent()) != 0 {
+			// Need to clone
+			if err := d.clone(volumeID, source.GetParent()); err != nil {
+				d.Delete(v.GetId())
+				return "", err
+			}
+		}
 
-	if err := d.CreateVol(v); err != nil {
-		return "", err
+		// Set to ready
+		v.State = api.VolumeState_VOLUME_STATE_AVAILABLE
+		if err := d.UpdateVol(v); err != nil {
+			deleteErr := d.Delete(v.GetId())
+			if deleteErr != nil {
+				logrus.Errorf("Failed to clean up and delete volume %s: %v", volumeID, deleteErr)
+			}
+			logrus.Errorf("Failed to update volume %s to ready state: %v", volumeID, err)
+			return "", err
+		}
+	} else {
+		// Block volume
+		if spec.GetSize() == 0 {
+			return "", fmt.Errorf("Cannot have size of zero on block volume")
+		}
+		v = common.NewVolume(
+			volumeID,
+			spec.GetFormat(),
+			locator,
+			source,
+			spec,
+		)
+
+		// Set as pending until the volume is ready
+		v.State = api.VolumeState_VOLUME_STATE_PENDING
+		if err := d.CreateVol(v); err != nil {
+			return "", err
+		}
+
+		// Check if this from as a clone
+		if source != nil && len(source.GetParent()) != 0 {
+			// Need to clone
+			if err := d.clone(volumeID, source.GetParent()); err != nil {
+				deleteErr := d.Delete(v.GetId())
+				if deleteErr != nil {
+					logrus.Errorf("Failed to clean up and delete volume %s: %v", volumeID, deleteErr)
+				}
+				return "", err
+			}
+		} else {
+			// This is a new volume
+			blockFile := path.Join(volPathParent, volumeID+nfsBlockFile)
+			f, err := os.Create(blockFile)
+			if err != nil {
+				logrus.Errorf("Unable to create block file %s: %v", blockFile, err)
+				return "", err
+			}
+			defer f.Close()
+
+			// Create sparse file
+			if err := f.Truncate(int64(spec.Size)); err != nil {
+				logrus.Println(err)
+				return "", err
+			}
+
+			// Format
+			if spec.GetFormat() != api.FSType_FS_TYPE_NONE {
+				dev, err := losetup.Attach(blockFile, 0, false)
+				if err != nil {
+					return "", err
+				}
+				defer func() {
+					if err := dev.Detach(); err != nil {
+						logrus.Errorf("Failed to detach %s", dev)
+					}
+				}()
+				logrus.Infof("Formatting %s with %v", dev, spec.Format)
+
+				// Get mkfs
+				cmd, err := mkfsFormatTypeCmd(spec.Format.SimpleString())
```

Comment:
I thought we decide to make Mount operation as no-op for block devices. So we will never format the volume and treat a volume with FsType set as a regular volume.

Reply:
For Raw, yes. For FS block volumes, we need to mount to format.